trunk/sf/sf.3
.\"/*+
.\"***********************************************************************
.\"***  C A N A D I A N   A S T R O N O M Y   D A T A   C E N T R E  *****
.\"
.\" (c) <year>				(c) <year>
.\" National Research Council		Conseil national de recherches
.\" Ottawa, Canada, K1A 0R6 		Ottawa, Canada, K1A 0R6
.\" All rights reserved			Tous droits reserves
.\" 					
.\" NRC disclaims any warranties,	Le CNRC denie toute garantie
.\" expressed, implied, or statu-	enoncee, implicite ou legale,
.\" tory, of any kind with respect	de quelque nature que se soit,
.\" to the software, including		concernant le logiciel, y com-
.\" without limitation any war-		pris sans restriction toute
.\" ranty of merchantability or		garantie de valeur marchande
.\" fitness for a particular pur-	ou de pertinence pour un usage
.\" pose.  NRC shall not be liable	particulier.  Le CNRC ne
.\" in any event for any damages,	pourra en aucun cas etre tenu
.\" whether direct or indirect,		responsable de tout dommage,
.\" special or general, consequen-	direct ou indirect, particul-
.\" tial or incidental, arising		ier ou general, accessoire ou
.\" from the use of the software.	fortuit, resultant de l'utili-
.\" 					sation du logiciel.
.\"
.\"***********************************************************************
.\"
.\" FILENAME
.\" sf.3
.\"
.\" PURPOSE:
.\" Describe the sds to fits library.
.\"
.\"
.\"INDENT-OFF*
.\" $Log: not supported by cvs2svn $
.\" Revision 1.1.1.1  2002/02/21 20:23:34  tpaz
.\"
.\"
.\" Revision 1.1.1.1  1999/08/23 19:58:05  cvs-tuc
.\" Initial install into CVS of dhs-0.16
.\"
.\" Revision 1.2  1997/10/29 23:52:40  jaeger
.\" removed all dd duplication, added macros for formating, updated to
.\" correct information.
.\"
.\" Revision 1.1  1997/10/10 21:02:04  dunn
.\" Initial revision
.\"
.\"INDENT-ON*
.\"
.\"***  C A N A D I A N   A S T R O N O M Y   D A T A   C E N T R E  *****
.\"***********************************************************************
.\"-*/
.\"
.                       \" Ty macro to start an indent for each type def.
.de Ty
.ie n .TP 14n
.el .TP 16n
..
./"
.                       \" Ss macro to start a structure definition.
.de Ss
.PP
.RS
.nf
.ft 3
.ie n .ta 12n 26n 55n
.el .ta 16n 40n 70n
..
./"
.                       \" Es macro to end a structure definition.
.\"
.de Es
.ft 1
.fi
.RE
.PP
..
.\"
.TH sf 3 28/10/97 "CADC"


.SH NAME
sf, sfCheckWithDd, sfCloseFitsFile, sfDsInfoFree, sfDsInfoInit, sfExit, sfExtractKeyword, sfFitsFileSize, sfFitsFlush, sfFits2Sds, sfGetValue, sfGetFilename, sfInit, sfObjectFree, sfOpenFitsFile, sfPrintMessages, sfSds2Fits, sfSetFitsMem, sfWriteDhsObject, sfWriteMemToFile
\- Library to convert between the Self-defining Data Structure (SDS) and 
the Flexible Image Transport System (FITS) file formats.


.SH SYNOPSIS
.LP
.nf
#include "gen_types.h"		
#include "gen_eptr.h"
#include "gen_str.h"
#include "gen_msg.h"

#include "sf.h
.ft
.fi
.LP
.\".nf
.\".ta 0.8i
.TP 11n
void
sfCheckWithDd( char[FLEN_KEYWORD]
.IR instrument,
DHS_DATA_TYPE
.IR type,
char *
.IR attribute,
DD_ARRAY_TYPE
.IR array,
char[FLEN_COMMENT]
.IR comm,
char[FLEN_KEYWORD]
.IR keyword,
boolean
.IR writeToFile,
SF_STATUS *
.IR sfStatus
);
.TP 11n
void
sfCloseFitsFile( fitsfile *
.IR fptr,
SF_STATUS *
.IR sfStatus
);
.TP 11n
void
sfDsInfoFree( SF_DS_INFO *
.IR dsInfo
);
.TP 11n
SF_DS_INFO 
*sfDsInfoInit( const char *
.IR name,
SF_STATUS *
.IR sfStatus
);
.TP 11n
void 
sfExit( SF_STATUS *
.IR sfStatus
);
.TP 11n
char 
*sfExtractKeyword( char
.IR card,
SF_STATUS *
.IR sfStatus
);
.TP 11n
long    
sfFitsFileSize( fitsfile *
.IR fptr,
SF_STATUS * 
.IR sfStatus
);
.TP 11n
void        
sfFitsFlush( fitsfile *
.IR fptr,
SF_STATUS * 
.IR sfStatus
);
.TP 11n
void        
sfFits2Sds( char *
.IR fitsFileName,
int
.IR numXAxisDiv,
int
.IR numYAxisDiv,
boolean
.IR checkDd,
boolean 
.IR ddMake
boolean
.IR addArrays,
EPTR, 
.IR objectList,
SF_STATUS * 
.IR sfStatus
);
.TP 11n
void        
sfGetValue( DHS_BD_OBJECT
.IR dhsObject,
char * 
.IR name,
DHS_DATA_TYPE * 
.IR type,
int * 
.IR ndims,
unsigned long * 
.IR dims,
void ** 
.IR value, 
SF_STATUS *
.IR sfStatus 
);
.TP 11n
char        
*sfGetFilename( const char *
.IR pathName,
const char * 
.IR frameId,
const char *
.IR ext,
SF_STATUS *
.IR sfStatus
);
.TP 11n
void 
sfInit( SF_STATUS *
.IR sfStatus
);
.TP 11n
void
sfObjectFree(  SF_STATUS *
.IR sfStatus
);
.TP 11n
fitsfile    
*sfOpenFitsFile( char *
.IR fitsFileName, 
char 
.IR ReadWrite,
SF_STATUS *
.IR sfStatus
);
.TP 11n
void        
sfPrintMessages( char *
.IR message,
int
.IR type
);
.TP 11n
void        
sfSds2Fits( DHS_BD_OBJECT
.IR sdsObject,
fitsfile * 
.IR fptr,
SF_DS_INFO * 
.IR datasetInfo,
boolean
.IR oneChunk,
SF_STATUS *
.IR sfStatus
);
.TP 11n
fitsfile    
*sfSetFitsMem( void ** 
.IR buffer,
long * 
.IR size, 
SF_DS_INFO *  
.IR dsInfo, 
SF_STATUS *
.IR sfStatus
);
.TP 11n
void        
sfWriteDhsObject( DHS_BD_OBJECT 
.IR dhsObject,
const char * 
.IR pathName, 
const char *
.IR objectId, 
const char * 
.IR ext,
boolean 
.IR oneChunk, 
SF_DS_INFO * 
.IR datasetInfo,
SF_STATUS *
.IR sfStatus 
);
.TP 11n
void        
sfWriteMemToFile( void ** 
.IR buffer, 
char * 
.IR fileName,
long * 
.IR size,
SF_STATUS *
.IR sfStatus
);


.SH TYPEDEFS
The following datatypes are defined in the sf.h header file.
.Ty
NAME_SIZE
This is the number of characters allowed in dataset names and 
extension identifiers.  This value should be the same as
.B SDS_NAME_SIZE
 in the 
.B DHS
library.
.Ty
NULL_INDEX
This is the NULL index into a FITS file.
.Ty
SF_DS_INFO
This structure holds information on a given dataset.
.Ss
char    name[NAME_SIZE];       /\(** Name of the dataset.             \(**/
int     numKeywords;           /\(** Number of keywords in Primary.   \(**/
                               /\(** FITS Header unit.                \(**/
char    instrument[FLEN_VALUE];/\(** The Instrument for the dataset.  \(**/
int     last;                  /\(** Index to last HDU in a FITS file.\(**/
EPTR    *exts;                 /\(**  List of extension information.  \(**/
.Es
.Ty
SF_EXT_INFO
This structure holds ths information about a particular FITS extension
that is created for each unique frame.
.Ss
char    name[FLEN_VALUE];       /\(** Value of "EXTNAME" keyword.    \(**/
char    *frmName;               /\(** Value of "FRMNAME" keyword.    \(**/
int     numKeywords;            /\(** Number of keywords in header.  \(**/
int     fitsIndex;              /\(** Index into the FITS file.      \(**/
SF_XTENSION type;               /\(** Extensions type, IMAGE, BINARY.\(**/
int     naxis;                  /\(** Number of axies.               \(**/
long    naxes[SF_MAXDIM];       /\(** Size of each axis.             \(**/
.Es
.Ty
SF_MAXIMUM
This is the maximum number of dimensions allowed for array and image data.
.Ty
SF_NULL_SIZE
This indicates a FITS file of NULL size.
.Ty
SF_OBJECT
This is a useful structure for creating test data.  It contains a
DHS dataset identifier and a unique string to identify it, the unique
string can be a filename.
.Ss
DHS_BD_OBJECT dhsObject;    /\(** A DHS BD object.            \(**/
char          *id;          /\(** Unique dataset identifier.  \(**/
.Es

.SH DESCRIPTION
.B sf
library provides utilites for converting between the Self-defining 
Data Structure and the Flexible Image Tranport System 
.B (FITS).

The 
.I cfitsio
library is used to manipulate the FITS files.  See the CFITSIO User's
Guide for more information.

The DHS bulk data facilities described in ICD3 are used to manipulate
the SDS structures.

Following is a list of the library's functions:
.Ty
sfCheckWithDd()
This function checks the validity of an attribute with the data dictionary.
It also determines whether the attribute is a multi-dimensional array or
not, the FITS keyword and the comments to be written to the FITS file.  
.Ty
sfCloseFitsFile()
This function closes a FITS file pointer.
.Ty
sfDsInfoFree()
This function frees all of the memory used by a dataset information structure
(SF_DS_INFO).
.Ty
sfDsInfoInit()
This function creates a new dataset information structure and returns it.  A NULL is returned if an error occurs.  Memory is allocated for the dataset information.
.Ty
sfExit()
This function cleans up the sf library before exiting.
.Ty
sfExtractKeyword()
This function extracts a FITS keyword from a card read from a FITS file.
.Ty
sfFitsFileSize()
This function determines the file size, in bytes, of a  FITS file.
.Ty
sfFitsFlush()
This function causes any internal data to be flushed in to the FITS
output file.  This is particularly useful with FITS memory buffers,
it forces the FITS file to be in the specified memory location.
.Ty
sfFits2Sds()
This function converts a FITS file into a SDS object. The SDS structure(s)
that are generated by this fucntion are intended to be used as test
data for the DHS system.

The image data in the FITS file 
is broken up into "chunks".  The number of chunks is determined by the
.IR numxc
and
.IR numyc
parameters.  More specifically, the
.IR numxc
parameter indicates the number of divisions along the x-axis and 
.IR numyc
parameter indicated the number of divisions along the y-axis.

If the 
.IR checkDd
parameter is true then the attributes in the SDS structure are determined
from the data dictionary.

If the 
.IR ddMake
parameter is true then a data dictionary file will be generated and the
SDS attributes are the same as the FITS keywords.

If the
.IR array
parameter is true then multi-dimensional arrays are randomly generated
and added to the SDS structure.
.Ty
sfGetValue()
This function determines finds the value of a SDS attribute.
.Ty
sfGetFilename()
This function creates generates a file name for a FITS file.  It is intended
to be used when converting from FITS files to SDS structures.  The new file
name is based on the input files path name and the frame identifier.  The
new path name will be of the form :

.I baseName.frameId.extension

where basenName is the file name of the original file without its extension,
frameId is the frame identifier, and extension is the file extension.
.Ty
sfInit()
This function initializes the sf library.  It also initializes the
data dictionary.
.Ty
sfObjectFree()
This function frees a SF object from memory when used with the 
.IR eptr_free
function.
See 
.IR gen_etpr
for more details on
IR eptr_free.
.Ty
sfOpenFitsFile()
This function opens a FITS file for reading and/or writing.
.Ty
sfPrintMessages()
Prints the "message" to the log files and, depending on the type,
to stdout.  The message is then cleared.
.Ty
sfSds2Fits()
This function converts a SDS structres sent to DHS from instruments to 
a FITS file. The
.IR oneChunk
parameter indicates whether the SDS structure is self-contained (complete)
or if it is a "chunk".  Multiple SDS chunks can be composed into a
single FITS file.  When SDS chunks are composed the
.IR dsInfo
parameter is expected to be filled in, i.e. it describes the SDS
structure.  This can be constructed by converting each SDS chunk
with 
.IR oneChunk
set to "TRUE"  before composing.
.Ty
sfSetFitsMem()
This function initializes a memory buffer that will hold a FITS file.
.Ty
sfWriteDhsObject()
This function writes a SDS structure to disk.  The file is written in
FITS format if the value of the
.IR ext
parameter is "fit". If the value of the
.IR ext
parameter is "sds" then an SDS buffer is written to the file.
.Ty
sfWriteMemToFile()
This function writes a FITS memory buffer to a disk file.


.SH FILES

.SS

.B sf
also requires a configuration file called
.I libdd.config
which contains the data dictionary. See the dd man page for 
specifications.

.SH ENVIRONMENT VARIABLES

The environment variable
.B LIBDD_CONFIG_DIR may be set to the path of the 
directory where the 
.B sf
configuration file can be found. If
.B LIBDD_CONFIG_DIR
is not set, the environment variable
.B DEFAULT_CONFIG_DIR 
should be set to the path of the directory where the file can be found. 


.SH DIAGNOSTICS

none.


.SH SEE ALSO

.BR dhsData (1).
.BR sf (3).
.BR fh (3).


.SH BUGS
none.

.SH AUTHORS
Jennifer Dunn
.br
Shannon Jaeger

.br

National Research Council
.br
Herzberg Institute of Astrophysics
.br
Canadian Astronomy Data Centre




CHANGES
==============================================
   DHS-lite sybase-less cfitio3.0.6 release
==============================================

gen++/genListOf.H                disable copy constructor
dhsStatus/staChannel.C           fix definition
dhsStatus/staChannel.H           fix missing public statement
dhsData/msg.C                    fix typo
dhsGen/dhsGenTmp.C               fix read_dir
dhsData/putFits.C                implement sybase-less option
dhsCommand/cmdCommandInfo.C      fix iterator declaration
dhsCommand/cmdDhs.H              fix friend declaration
conf/cxx_lib.mk                  fix clean target
conf/cc_bin.mk                   fix clean target
conf/cxx_bin.mk                  fix clean target
conf/cc_lib.mk                   fix clean target
conf/tcl.mk                      fix clean target
press/press.c                    add missing header
fh/fh.C                          precompiler switch for outdated cfitsio
dhsQlServer/qlsFrame.C           precompiler switch for outdated cfitsio
dhsQlServer/qlsDataset.C         precompiler switch for outdated cfitsio
sf/utils.c                       fix chunk size calculation, precompiler switch for outdated cfitsio
qldp/image.C                     fix Fits_IO.h header name under skycat-3.0.1
qldp/image.H                     fix Fits_IO.h header name under skycat-3.0.1
cdIngest/ingest.c                fix weird declaration?
sf/sds2fits.c                    leave libdd.config translation bug (would break GSA etc.)
                                 catch any attempt to move to an invalid hdu if maxHdr==0
                                 enforce max keys length to avoid HIERARCH convention
                                 precompiler switch for outdated cfitsio
conf/include.mk.in               several changes to handle sybase-less dhs, outdate-cfitsio,
                                 standard location for DRAMA, EPICS, etc
configure.in                     several changes to handle sybase-less dhs, outdate-cfitsio,
                                 standard location for DRAMA, EPICS, etc
dhsData/putSds.C                 critical fix for mem mapped fits
                                 precompiler switch for sybase-less dhs
dhsStatus/staMonitor.H           precompiler switch for sybase-less
dhsStatus/staMonitor.C           precompiler switch for sybase-less
dhsData/dtsDhs.C                 precompiler switch for sybase-less
dhsData/get.H                    precompiler switch for sybase-less
dhsData/get.C                    precompiler switch for sybase-less (needs improvement)
dhsStatus/Makefile               confitional flag for sybase-less
dhsData/Makefile                 confitional flag for sybase-less
dhsData/putFits.C                precompiler switch for sybase-less
dhsData/msg.C                    removing ending .
dhsData/dbm.C                    precompiler switch for sybase-less



trunk/sf/sf.c
static char rcsid[] = "$Id: sf.c,v 1.1.1.1 2002-11-24 20:35:47 brighton Exp $";
/*
************************************************************************
****  C A N A D I A N	A S T R O N O M Y   D A T A   C E N T R E  *****
*
* (c) <year>				(c) <year>
* National Research Council		Conseil national de recherches
* Ottawa, Canada, K1A 0R6		Ottawa, Canada, K1A 0R6
* All rights reserved			Tous droits reserves
*					
* NRC disclaims any warranties,		Le CNRC denie toute garantie
* expressed, implied, or statu-		enoncee, implicite ou legale,
* tory, of any kind with respect	de quelque nature que se soit,
* to the software, including		concernant le logiciel, y com-
* without limitation any war-		pris sans restriction toute
* ranty of merchantability or		garantie de valeur marchande
* fitness for a particular pur-		ou de pertinence pour un usage
* pose.	 NRC shall not be liable	particulier.  Le CNRC ne
* in any event for any damages,		pourra en aucun cas etre tenu
* whether direct or indirect,		responsable de tout dommage,
* special or general, consequen-	direct ou indirect, particul-
* tial or incidental, arising		ier ou general, accessoire ou
* from the use of the software.		fortuit, resultant de l'utili-
*					sation du logiciel.
 *
 ************************************************************************
 *
 * FILENAME
 * sf/src/sf.c
 *
 * PURPOSE:
 * Library to convert between FITS and SDS file formats.

 *
 * FUNCTION NAME(S)
 * in	    sfExit	: Cleanup the SF library.
 * int	    sfInit	: Initialize SF library.
 * void	    sfInitSet   : Set the SF library initialization status.
 * boolean  sfInitTest	: Check SF initialization status.
 *
 *INDENT-OFF*
 * $Log: not supported by cvs2svn $
 * Revision 1.1.1.1  2002/02/21 20:23:34  tpaz
 *
 *
 * Revision 1.1.1.1  1999/08/23 19:58:04  cvs-tuc
 * Initial install into CVS of dhs-0.16
 *
 * Revision 1.9  1997/08/22 22:10:35  jaeger
 * Added support for multi-dimensional arrays.
 *
 * Revision 1.8  1997/08/15 20:43:37  jaeger
 * Checked in for install
 *
 * Revision 1.7  1997/08/12 21:07:38  jaeger
 * Checked in for install
 *
 * Revision 1.6  1997/08/06 18:30:48  jaeger
 *  Checked in for install and testing with dhs
 *
 * Revision 1.5	 1997/07/09 16:32:45  jaeger
 *  Converted library from bdt to dhs
 *
 *INDENT-ON*
 *
 ****  C A N A D I A N	 A S T R O N O M Y   D A T A   C E N T R E  *****
 ************************************************************************
*/

/*
 ************************************************************************
 *  CADC Library include files.
 ************************************************************************
 */

#include "gen_types.h"
#include "gen_eptr.h"
#include "gen_msg.h"

/*
 ************************************************************************
 *  cfitsio library include files.
 ************************************************************************
 */

#include "fitsio.h"


/*
 ************************************************************************
 *  sf library include files.
 ************************************************************************
 */

#include "sf.h"
#include "localSf.h"


static 	boolean	initialized	= FALSE;

void	sfInitSet( boolean );

void	sfFormatMessage();



/*
 ************************************************************************
 *+
 * FUNCTION NAME:
 * sfExit
 *
 * INVOCATION:
 * sfExit( sfStatus );
 *
 * PARAMETERS: (">" input, "!" modified, "<" output)
 * (>) sfStatus 	(SF_STATUS*) 	SF library function status.	
 *
 * FUNCTION VALUE:
 * None
 *
 * PURPOSE:
 * Cleanup the library before exiting.
 *
 * DESCRIPTION:
 * Description of the algorithm.
 *
 * EXTERNAL VARIABLES:
 * None.
 *
 * PRIOR REQUIREMENTS:
 * None.
 *
 * SEE ALSO:
 * - sfInit.
 *
 * DEFICIENCIES:
 * None.
 *-
 ************************************************************************
 */


void sfExit
( 
    SF_STATUS	*sfStatus	/* (mod) Function return status.	*/
)
{
    DD_STATUS	ddStatus;
    
    fnEntry ( "sfExit", *sfStatus );
    statCheck( *sfStatus,  VOID );

    ddStatus = DD_S_SUCCESS;
    ddCheck( ddExit( &ddStatus ), ddStatus, *sfStatus, VOID );
    fnReturn( VOID );
}



/*
 ************************************************************************
 *+
 * FUNCTION NAME:
 * sfInit
 *
 * INVOCATION:
 * sfInit ( sfStatus );
 *
 * PARAMETERS: (">" input, "!" modified, "<" output)
 * (>) sfStatus 	(SF_STATUS*) 	SF library function status.	
 *
 * FUNCTION VALUE:
 * None
 *
 * PURPOSE:
 * Initialize the data dictionary.
 *
 * DESCRIPTION:
 * Call ddInit to intialize the data dictionary.
 * Append any messages from dd, that occured during intialization
 * to the SF messages.
 *
 * EXTERNAL VARIABLES:
 * None.
 *
 * PRIOR REQUIREMENTS:
 * None.
 *
 * SEE ALSO:
 * - sfExit
 *
 * DEFICIENCIES:
 * None.
 *-
 ************************************************************************
 */

void sfInit
( 
    SF_STATUS	*sfStatus	/* (mod) Function return status.	*/
)
{
    DD_STATUS	ddStatus;
    
    fnEntry ( "sfInit", *sfStatus );
    statCheck( *sfStatus,  VOID );

    ddStatus = DD_S_SUCCESS;
    ddCheck( ddInit( &ddStatus ), ddStatus, *sfStatus, VOID );
    msg_append( sfMsg, ddMsg );
    msg_clear( ddMsg );
    

    sfInitSet( TRUE );
        
    fnReturn( VOID );
}



/*
 ************************************************************************
 *+
 * FUNCTION NAME:
 * sfInitSet
 *
 * INVOCATION:
 * sfInitSet( booleanValue )
 *
 * PARAMETERS: (">" input, "!" modified, "<" output)
 * (>) value	(boolean)  Initialization value (TRUE/FALSE).
 *
 * FUNCTION VALUE:
 * None.
 *
 * PURPOSE:
 * Set the initialized state of the SF library
 *
 * DESCRIPTION:
 * Trivial.
 *
 * EXTERNAL VARIABLES:
 * "initialized" - Keeps track of the current state.
 *
 * PRIOR REQUIREMENTS:
 * None.
 *
 * SEE ALSO:
 * None..
 *
 * DEFICIENCIES:
 * None.
 *-
 ************************************************************************
 */

void sfInitSet
(
    boolean	value		/* (in)  Initialization value.		*/
)
{
    initialized = value;
}



/*
 ************************************************************************
 *+
 * FUNCTION NAME:
 * sfInitTest
 *
 * INVOCATION:
 * booleanValue = sfInitTest( )
 *
 * PARAMETERS: (">" input, "!" modified, "<" output)
 * None.
 *
 * FUNCTION VALUE:
 * (boolean) TRUE  : If "sfInit" has been executed.
 * (boolean) FALSE : If "sfInit" has not been executed.
 *
 * PURPOSE:
 * Return the current initialization state of the SF library.
 *
 * DESCRIPTION:
 * Trivial.
 *
 * EXTERNAL VARIABLES:
 * "initialized" - Keeps track of the current state.
 *
 * PRIOR REQUIREMENTS:
 * None.
 *
 * SEE ALSO:
 * None.
 *
 * DEFICIENCIES:
 * None.
 *-
 ************************************************************************
 */

boolean	sfInitTest
(
    void
)
{
    return( initialized );
}
